src/coprocessor/codec/batch/rows.rs

```diff
@@ -0,0 +1,716 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO: Remove this.
+#![allow(dead_code)]
+
+use std::convert::TryFrom;
+
+use smallvec::SmallVec;
+
+use cop_datatype::{EvalType, FieldTypeAccessor, FieldTypeFlag};
+use tipb::schema::ColumnInfo;
+
+use super::BatchColumn;
+use coprocessor::codec::mysql::Tz;
+use coprocessor::codec::{datum, Error, Result};
+
+pub struct BatchRows<E> {
+    /// Multiple interested columns. Each column is either decoded, or not decoded.
+    ///
+    /// For decoded columns, they may be in different types. If the column is in
+    /// type `LazyBatchColumn::Encoded`, it means that it is not decoded.
+    columns: Vec<LazyBatchColumn>,
+
+    /// The number of rows (either encoded or decoded) stored.
+    number_of_rows: usize,
+
+    /// The optional error occurred during batch execution. Once there is an error, this field
+    /// is set and no more rows should be fetched or executed.
+    some_error: Option<E>,
+}
+
+impl<E: Clone> Clone for BatchRows<E> {
+    fn clone(&self) -> Self {
+        Self {
+            columns: self.columns.clone(),
+            number_of_rows: self.number_of_rows,
+            some_error: self.some_error.clone(),
+        }
+    }
+}
+
+impl<E> BatchRows<E> {
+    /// Creates a new `BatchRows`, which contains `columns_count` number of raw columns
+    /// and each of them reserves capacity according to `rows_capacity`.
+    pub fn raw(columns_count: usize, rows_capacity: usize) -> Self {
```

Comment:
Is "raw_with_capacity" better?

Reply:
I'm fine with both. The function accepts both columns and rows but it would be too long to include both of them in the name 🤣So I decided to ignore them both.

Reply:
But using a function named `raw` to new an object looks weird.